File: go/300.500/473.MatchsticksToSquare.go
package main

import (
	"fmt"
	"sort"
)

func makesquare(matchsticks []int) bool {
	sum := 0
	for _, x := range matchsticks {
		sum += x
	}
	if sum%4 != 0 {
		return false
	}
	sort.Slice(matchsticks, func(i, j int) bool {
		return matchsticks[i] > matchsticks[j]
	})
	elen := sum / 4
	var dfs func(index int) bool
	edges := make([]int, 4)
	size := len(matchsticks)
	dfs = func(index int) bool {
		if index == size {
			return true
		}
		for i := range edges {
			edges[i] += matchsticks[index]
			if edges[i] <= elen {
				if dfs(index + 1) {
					return true
				}
			}
			edges[i] -= matchsticks[index]
		}
		return false
	}
	return dfs(0)
}

func makesquare1(matchsticks []int) bool {
	sum := 0
	for _, x := range matchsticks {
		sum += x
	}
	if sum%4 != 0 {
		return false
	}
	elen := sum / 4
	dp := make([]int, 1<<len(matchsticks))
	for i := range dp {
		dp[i] = -1
	}
	dp[0] = 0
	for s := 1; s < len(dp); s++ {
		for k, v := range matchsticks {
			if s>>k&1 == 0 {
				continue
			}
			s1 := s &^ (1 << k)
			if dp[s1] >= 0 && dp[s1]+v <= elen {
				dp[s] = (dp[s1] + v) % elen
				break
			}
		}
	}
	return dp[len(dp)-1] == 0
}

func main() {
	fmt.Println(makesquare1([]int{1, 1, 2, 2, 2}), true)
	fmt.Println(makesquare1([]int{3, 3, 3, 3, 4}), false)
}

File: 300.500/473.MatchsticksToSquare.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <algorithm>
using namespace std;

class MatchsticksToSquare{
public:
    bool makesquare(vector<int> matchsticks) {
        int sum = accumulate(matchsticks.begin(), matchsticks.end(), 0);
        if (sum % 4 != 0){
            return false;
        }
        sort(matchsticks.begin(), matchsticks.end(), greater<int>());
        vector<int> edges(4);
        return dfs(0, matchsticks, edges, sum / 4);
    }
    bool dfs(int index, vector<int>& matchsticks, vector<int>& edges, int len){
        if(index == matchsticks.size()){
            return true;
        }
        for(int i =0;i<edges.size();++i){
            edges[i] += matchsticks[index];
            if(edges[i] <= len && dfs(index+1,matchsticks, edges, len)){
                return true;
            }
            edges[i] -= matchsticks[index];
        }
        return false;
    }
    bool makesquare1(vector<int> matchsticks) {
        int sum = accumulate(matchsticks.begin(), matchsticks.end(), 0);
        if (sum % 4 != 0){
            return false;
        }
        int edge = sum / 4;
        int size = matchsticks.size();
        vector<int> dp(1<<size, -1);
        dp[0]=0;
        for(int s=1;s<(1<<size);++s){
            for(int k=0;k<size;++k){
                if(s & (1<<k) == 0){    //这里表示第k根已经放进去了 所以不重复放
                    continue;
                }
                int s1 = s & ~(1 << k);  //s1标识s去掉第k根
                if(dp[s1] >= 0 && dp[s1] + matchsticks[k] <= edge){
                    dp[s] = (dp[s1] + matchsticks[k]) % edge;
                    break;
                }
            }
        }
        return dp[(1<<size) - 1] == 0;
    }
};

int main(){
    MatchsticksToSquare cls;
    cout << boolalpha << cls.makesquare1(vector<int>{1,1,2,2,2}) << " -> true" << endl;
    cout << boolalpha << cls.makesquare1(vector<int>{3,3,3,3,4}) << " -> false" << endl;
    return 0;
}
